pipeline/matching/core/analytics_test.py

```diff
@@ -0,0 +1,174 @@
+from unittest.mock import MagicMock
+
+import pytest
+
+from pipeline.matching.core.analytics import CollectInternalAnalytics
+from pipeline.matching.core.engine import MatchingEngine
+from pipeline.types import Match, MatchingInput, MatchingInternalAnalytics, User
+
+
+def test_analytics_logic():
+
+    users = [
+        User(uid="A", displayName="User A"),
+        User(uid="B", displayName="User B"),
+        User(uid="C", displayName="User C"),
+        User(uid="D", displayName="User D"),
+        User(uid="E", displayName="User E"),
+    ]
+    recent_matches = [Match(users={"A", "B"}), Match(users={"C", "D", "E"})]
+    inp = MatchingInput(
+        community="test",
+        release="2022-04-01",
+        users=users,
+        recent_matches=recent_matches,
+    )
+    mock_generator = MagicMock()
+    mock_generator.generate = MagicMock(
+        return_value=[
+            Match(users={"A", "B"}),
+            Match(users={"C", "D"}),
+        ]
+    )
+    mock_ranker = MagicMock()
+    mock_ranker.rank = MagicMock(return_value=[Match(users={"A", "D"})])
```

Comment:
Good job on learning how to use `MagicMock`!

We shouldn't need all this setup, all we need is one iterator of matches as an input, then these tests will be much shorter and easier to understand.

An iterator can be a list, but if you want to really match the input the engine will give you, we should use a generator. You can turn a list into a generator by calling the `iter()` method on it.

For example, this is what your test bodies will look like, much simpler:

```python
matches = [
    Match(users={"A", "B"}),
    Match(users={"C", "D"}),
]
analyzer = CollectInternalAnalytics()
actual = analyzer.n_proposed_matches_per_user(iter(matches))
...
assert actual == expected
```

Reply:
Wow...that is very simple. Good to know. Just another instance of me slightly overcomplicating the objective. I will revise my test cases accordingly.

Reply:
Excellent, great job refactoring the tests!